extensions/amp-ad-network-smartadserver-impl/0.1/amp-ad-network-smartadserver-impl.js

```diff
@@ -0,0 +1,206 @@
+import {tryParseJson} from '#core/types/object/json';
+import {includes} from '#core/types/string';
+
+import {Services} from '#service';
+
+import {dev} from '#utils/log';
+
+import {getOrCreateAdCid} from '../../../src/ad-cid';
+import {getConsentPolicyInfo} from '../../../src/consent';
+import {AmpA4A} from '../../amp-a4a/0.1/amp-a4a';
+
+/** @type {string} */
+const TAG = 'amp-ad-network-smartadserver-impl';
+
+/** @const {number} */
+const MAX_URL_LENGTH = 15360;
+
+/** @type {string} */
+const SAS_NO_AD_STR = '<html><head></head><body></body></html>';
+
+/**
+ * @const {!./shared/url-builder.QueryParameterDef}
+ * @visibleForTesting
+ */
+const TRUNCATION_PARAM = {
+  name: 'trunc',
+  value: 1,
+};
+
+/** @final */
+export class AmpAdNetworkSmartadserverImpl extends AmpA4A {
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @protected {!Deferred<string>} */
+    this.getAdUrlDeferred = new Deferred();
+  }
+
+  /** @override */
+  getAdUrl(opt_consentTuple, opt_rtcResponsesPromise) {
+    Promise.any([
+      getConsentPolicyInfo(this.element, this.getConsentPolicy() || 'default'),
+      new Promise((resolve) => setTimeout(() => resolve(), 10)),
+    ]).then((consentString) => {
+      opt_rtcResponsesPromise = opt_rtcResponsesPromise || Promise.resolve();
+      const checkStillCurrent = this.verifyStillCurrent();
+
+      opt_rtcResponsesPromise.then((result) => {
+        checkStillCurrent();
+        const rtc = this.getBestRtcCallout_(result);
+        const urlParams = {};
+
+        if (rtc && Object.keys(rtc).length) {
+          urlParams['hb_bid'] = rtc.hb_bidder || '';
+          urlParams['hb_cpm'] = rtc.hb_pb;
+          urlParams['hb_ccy'] = 'USD';
+          urlParams['hb_cache_id'] = rtc.hb_cache_id || '';
+          urlParams['hb_cache_host'] = rtc.hb_cache_host || '';
+          urlParams['hb_cache_path'] = rtc.hb_cache_path || '';
+          urlParams['hb_width'] = this.element.getAttribute('width');
+          urlParams['hb_height'] = this.element.getAttribute('height');
+        }
+
+        const formatId = this.element.getAttribute('data-format');
+        const tagId = 'sas_' + formatId;
+        const adUrl = buildUrl(
+          (this.element.getAttribute('data-domain') ||
+            'https://www.smartadserver.com') + '/ac',
+          {
+            'siteid': this.element.getAttribute('data-site'),
+            'pgid': this.element.getAttribute('data-page'),
+            'fmtid': formatId,
+            'tgt': this.element.getAttribute('data-target'),
+            'tag': tagId,
+            'out': 'amp-hb',
+            ...urlParams,
+            'gdpr_consent': consentString,
+            'pgDomain': this.win.top.location.hostname,
+            'tmstp': Date.now(),
+          },
+          MAX_URL_LENGTH,
+          TRUNCATION_PARAM
+        );
+        this.getAdUrlDeferred.resolve(adUrl);
+      });
+    });
+    return this.getAdUrlDeferred.promise;
```

Comment:
Not getting the point of having a `Deferred` object here. Why not simply returning the `adUrl` in the chain promise?

Reply:
Simply returning a string (the `adUrl`) doesn't work - I've just tested to confirm. It's so because we're relying on promises before, I think. It's implemented the same way as in other fast-fetch extensions (e.g. the [doubleclick](https://github.com/ampproject/amphtml/blob/main/extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js) one).

Reply:
DoubleClick used `Deferred` for an entirely different reason - they are resolving their promise outside of `getAdUrl` call, and that is why it is needed. Here, no one else is reading `this.getAdUrlDeferred` at all, so I do not see why this Deferred object is needed. Can you elaborate what was not working when you tested? I'd be curious to know.

Reply:
You are right. Sorry, my bad. I've removed the `Deferred` object.

Reply:
Hello @powerivq, @calebcordry, I'm wondering what's the next step in this request. Do you have any other comments or are going to review the change more? Please share your feedback. Thanks.